src/hotspot/share/runtime/basicLock.inline.hpp

```diff
@@ -28,27 +28,27 @@
 #include "runtime/basicLock.hpp"
 
 inline markWord BasicLock::displaced_header() const {
-  assert(LockingMode == LM_LEGACY, "must be");
+  assert(!UseObjectMonitorTable, "must be");
   return markWord(get_metadata());
 }
 
 inline void BasicLock::set_displaced_header(markWord header) {
-  assert(LockingMode == LM_LEGACY, "must be");
+  assert(!UseObjectMonitorTable, "must be");
   Atomic::store(&_metadata, header.value());
 }
```

Comment:
So I was unsure of this.  It's only LM_LEGACY that needs the displaced header?  Lightweight locking doesn't use it even with the table?

Reply:
Correct, one of the main features with `LM_LIGHTWEIGHT` is that it does not need to displace the header when fast locking. It only displaces it when doing inflated locking, but then it is displaced in the ObjectMonitor*, not on the thread stack. (And with the table we also avoid this displacement)

`LM_LEGACY` uses this stack slot for two things. If it is 0 it means that it was recursively fast locked. Otherwise if it is the address which the markWord points to it is the displaced header which should be restored when unlocking. And when doing inflated locking a non-zero value is written so it does not look like a recursive fast lock.

Reply:
I got my displaced headers mixed up between ObjectMonitor and BasicLock.  Thanks for the explanation, I don't really know the legacy locking code that well.